Disable Granite.Code when Continue is enabled

Jazzcort opened the pull request:
## Description

With this feature, Granite.Code will check if Continue is enabled every 10 second. If Continue is enabled, a overlay will show up to block the entire webview of Granite.Code. The message on the overlay will inform users that Granite.Code is imcompatible with Continue and suggest users to disable Continue in order to continue using Granite.Code.
https://github.com/Granite-Code/granite-code/issues/99

## Checklist

- [ ] Design the overlay.
- [x] Ensure Granite.Code does not conflict with any enabled extensions. If a conflict is detected, automatically disable Granite.Code and display a message prompting the user to review and disable conflicting extensions.

## Screenshots


https://github.com/user-attachments/assets/fad27763-b6b3-45d3-a295-249ab2d5c603



## Testing instructions

If you want to have the same behavior as the demo video, you need to change "Continue.continue" to "redhat.granitecode" in `extensions/vscode/src/util/util.ts` for `incompatibleExtensionIds` and comment out "--disable-extension=redhat.granitecode" in `.vscode/launch.json` so you can test this feature with Granite.Code. Otherwise, this feature will block itself because it gets launched as Continue in the debug mode. When we merge this feature to continue-for-granite and launch from granite-code, it gets launched as Granite.code and it can then successfully detect Continue and disable itself.

Jazzcort committed:
Disable Granite.Code when Continue is enabled

With this feature, Granite.Code will check if any incompatible extensions
(e.g. Continue) is enabled when it first gets activated. Granite.Code
will also keep checking for every newly enabled extension. If these
extensions are enabled, a overlay will show up to block the entire
webview of Granite.Code. The message on the overlay will inform users
that Granite.Code is imcompatible with these extensions and suggest users
to disable them in order to continue using Granite.Code.
https://github.com/Granite-Code/granite-code/issues/99

owtaylor commented:
@allanday the exact sequence in the video might be a bit confusing because of the development setup, but does basic behavior seem OK? Is the message OK? (The fade-out is because Granite.Code might be activated before Continue and having the sidebar appear with the normal content then suddenly blink to the message looked bad.)

How do you see this interacting with the wizard?

 A) The wizard proceeds as normally, but when it gets to the end and the sidebar is shown, this behavior is there:
 B) There's a step zero of the setup wizard, where we ask the user to disable continue [relates to asking the user to disable copilot]

owtaylor commented on core/protocol/ideWebview.ts:
```diff
@@ -54,6 +54,7 @@ export type ToIdeFromWebviewProtocol = ToIdeFromWebviewOrCoreProtocol & {
     void,
   ];
   "edit/exit": [{ shouldFocusEditor: boolean }, void];
+  bumpWithExtensions: [undefined, boolean];
```
Maybe checkForIncompatibleExtensions

owtaylor commented on extensions/vscode/src/util/util.ts:
```diff
@@ -122,3 +122,16 @@ export function getExtensionVersion(): string {
+    }
+  }
+
+  return false;
```
Could be simplified to:

 return extensions.some(e => incompatibleExtensionIds.has(e.id))

But it doesn't quite make sense to me that the check here is for "incompatible extensions", but we display a fixed message about Continue. Maybe we make this function actually return the message to display, or undefined if there is no conflict?

owtaylor commented on gui/src/components/BumpOverlay.tsx:
```diff
@@ -0,0 +1,11 @@
+function BumpOverlay() {
```
Hmm, "Bump" here doesn't make sense to me. Maybe "IncompatibleExtensonsOverlay"?

fbricon commented:
instead of polling every 10s, have you tried registering an extensions change listener with vscode.extensions.onDidChange?
You could then only check extensions when necessary, and publish your event if needed.

Jazzcort commented:
@fbricon That's a better way of doing this! Thanks for the suggestion, I'll look into it!

fbricon commented:
see https://github.com/microsoft/vscode/issues/152805#issuecomment-1171683462

owtaylor commented on gui/src/components/Layout.tsx:
```diff
@@ -218,54 +232,75 @@ const Layout = () => {
   }, []);
 
+  useEffect(() => {
+    checkBump();
```
Maybe just define checkBump as a nested function here to simplify and avoid the useCallback?

owtaylor commented on gui/src/components/Layout.tsx:
```diff
@@ -218,54 +232,75 @@ const Layout = () => {
+    return () => {
+      clearInterval(intervalId);
+    };
+  }, []);
```
otherwise checkBump is needed as a dependency, right?

owtaylor commented on gui/src/components/Layout.tsx:
```diff
@@ -53,10 +56,21 @@ const Layout = () => {
   const showDialog = useAppSelector((state) => state.ui.showDialog);
 
+  const [bumpWithOtherExtensions, setBumpWithOtherExtensions] = useState(false);
+  const [bump, setBump] = useState(false);
```
Why the two variables?

Jazzcort replied:
I don't know what I was thinking back then. 😂 You're right!. We just need one.

owtaylor commented on gui/src/components/Layout.tsx:
```diff
@@ -218,54 +232,75 @@ const Layout = () => {
-              onEnter={() => {
-                dispatch(setShowDialog(false));
+    <>
+      {bumpWithOtherExtensions && <BumpOverlay />}
```
So, this is an issue we have when making changes for granite/customization - while we want "format on save" when we are working on our own code, or working on upstreamable changes, we don't want it on when working on a non-upstreamable changes to files like this, or we hard conflicts to resolve when I merge in upstream changes.

I might suggest temporarily turning off "format on save" when working on a file like this, and turning it back on afterwards. You can also save with "C-k s" but that's hard to remember.

Jazzcort replied:
Got it! I'll turn off "format on save" when it's the PR for granite/customization.

owtaylor reviewed (changes requested):
Generally looking good!

Jazzcort commented:
@owtaylor Since @fbricon suggested a more efficient way to receive information about extensions change. In this case, we can just setup a webviewListener at Layout.tsx to listen to the result of `bumpWithExtensions`. I'm going to change the implementation a bit and let you know when I'm done. Does it sound good? 😊

owtaylor commented:
> @owtaylor Since @fbricon suggested a more efficient way to receive information about extensions change. In this case, we can just setup a webviewListener at Layout.tsx to listen to the result of `bumpWithExtensions`. I'm going to change the implementation a bit and let you know when I'm done. Does it sound good? 😊

Sounds good!

allanday commented:
The behaviour in this PR seems good to me. The main thing I'd like to see changed is the presentation of the error placeholder.

In an ideal world, we would have a standard layout for these placeholders, which can be reused depending on the current state. The one that I have in the mockups looks like this:

![image](https://github.com/user-attachments/assets/a99f29a8-ce46-435d-af8c-004bcf634205)

If a standard layout like this isn't feasible then I would at least center the text and have a heading with a description text below. The text could be something like

```
**Incompatible Extension Enabled**

Granite.Code cannot be used while the Continue extension is enabled. Please disable Continue from the extension settings.
```

Jazzcort commented:
@allanday I'll start to work on the UI for this overlay tomorrow. Will post the draft here when I'm done so you can take a look. 😁

Jazzcort commented:
@allanday Here is the draft. Let me know if there are any changes needed! 😁

![Screenshot 2025-04-23 at 2 41 34 PM](https://github.com/user-attachments/assets/4f1237d2-29be-4944-8d01-8ee17025dd51)

owtaylor commented on extensions/vscode/src/granite/utils/compatibilityUtil.ts:
```diff
@@ -0,0 +1,38 @@
+
+export function setupExtensionCheck(webviewMessenger: VsCodeWebviewProtocol) {
+  // Check if incompatible extensions exist when Granite.Code is activated
+  // Keep sending the checking result until the response is received
```
OK, if I understand this approach, the way you do it is:
 - On startup, you send checkForIncompatibleExtensions every second to the message to the webview until it respond
 - You also send it when the set of incompatible extensions changes.

This is quite complicated, and while we *might* get lucky and get the message to the webview right after it sets up the listener, we might also miss that and not get the message for 1 second.

The previous approach where the webview sends the extension checkForIncompatibleExtensions on initialization and the extension sends updateIncompatibleExtensions a) in response to that b) if the set of installed extensions changes seems both easier to understand and more reliable.

owtaylor commented on extensions/vscode/src/granite/utils/compatibilityUtil.ts:
```diff
@@ -0,0 +1,38 @@
+    }
+  }, 1000);
+
+  // Resgister a listener to check compatibility for every newly enabled extension
```
Typo "Register" not "Resgister"

Jazzcort replied:
Thanks for the catch!

owtaylor commented on gui/src/components/Layout.tsx:
```diff
@@ -57,6 +58,8 @@ const Layout = () => {
 
   const showDialog = useAppSelector((state) => state.ui.showDialog);
 
+  const [bumpWithOtherExtensions, setBumpWithOtherExtensions] = useState(false);
```
Can we use "conflictsWithOtherExtensions" instead?

Jazzcort replied:
Yes, I'll change it!

owtaylor commented on gui/src/components/Layout.tsx:
```diff
@@ -221,51 +229,56 @@ const Layout = () => {
-                dispatch(setShowDialog(false));
+    <>
+      {bumpWithOtherExtensions && <IncompatibleExtensionsOverlay />}
+      <LocalStorageProvider>
```
For the future, I'd prefer this without reindenting, even if that results in weird/improper indentation.

```
+ <>
+ {bumpWithOtherExtensions && <IncompatibleExtensionsOverlay />}
  <LocalStorageProvider>
```

you can leave it like this for this PR, since I see there are major changes to Layout.tsx upstream that I'll need to resolve manually anyways.

Jazzcort committed:
Move the initial check to webview side

Instead of setting a interval to keep checking if the webview receives the result
of initial check for incompatible extensions, this commit make the webview to
take care of the initial check. This can enhance the user experience by
immediately disable Granite.Code rather than the slight delay due to the
interval.

owtaylor commented on extensions/vscode/src/granite/utils/compatibilityUtil.ts:
```diff
@@ -0,0 +1,19 @@
+  vscode.extensions.onDidChange(() => {
+    webviewMessenger.send(
+      "checkForIncompatibleExtensions",
+      checkForIncompatibleExtensions(),
```
I don't really like this as a "unsolicited reply" - it obviously works, but it makes things confusing to read. Just have two messages without a reply: checkForIncompatibleExtensions in one direction and updateIncompatibleExtensions in the other direction.

Jazzcort replied:
Got it! I'll change initial check sent from Layout.tsx to use ideMessenger.post as well, and all the checking result will be sent through updateIncompatibleExtensions.

owtaylor reviewed (changes requested):
Looks good - just a few small remaining issues!

owtaylor commented on extensions/vscode/src/granite/utils/compatibilityUtil.ts:
```diff
@@ -0,0 +1,19 @@
+
+export function setupExtensionCheck(webviewMessenger: VsCodeWebviewProtocol) {
+  // Register a listener to check compatibility for every newly enabled extension
+  vscode.extensions.onDidChange(() => {
```
I'm pretty sure that the right thing to do here is to pass the extensionContext into this function and do:
 `context.subscriptions.push(vscode.extensions.onDidChange(...)` - the vscode docs and examples do this quite consistently.

Note: There are quite a few places where our extension *doesn't* do this, so I'm a bit confused how it's both:
 - sort of OK to not do it (there must be some fallback mechanism to clean this up?)
 - Best practices to do it

Gemini claims: "Preventing Leaks: If you don't dispose of listeners, they might remain active even after your extension is supposed to be deactivated (especially in scenarios like extension reloading without a full window close). This can lead to unexpected behavior or memory leaks." Don't know if this is real or not.

In any case, I think we should stick to the documented patterns and unregister any newly added global listeners.

Jazzcort replied:
I have the same doubt as well. According to the documentation, `context.subscriptions.push(vscode.extensions.onDidChange(...)` is the right thing to do. It seems like during the deactivating process vscode will automatically call `.dispose()` on each  `Disposable` in this list. However, Continue doesn't implement `deactivate()` function so whenever users disable Continue, vscode provides sort of a defensive mechanism offering a Restart Extensions button for users to restart the entire extension system. I think those lingering event listeners, commands, and providers get disposed after this restart process.

I personally don't like this "Restart Extensions" thing because I think that in order to disable a extension users have to reactivate every single extension they enabled doesn't feel right. There are other extensions that implement `deactivate` function able to clean up themselves without the Restart Extensions mechanism.

I'll change the code to push the `Disposable` returned from`vscode.extensions.onDidChange(...)` to the `context.subscriptions` list

Jazzcort replied:
> However, Continue doesn't implement deactivate() function so whenever users disable Continue, vscode provides sort of a defensive mechanism offering a Restart Extensions button for users to restart the entire extension system.

I'm wrong about this. Continue does implement `deactivate` but I overlooked it. The "Restart Extensions" button appearing after disabling Continue is probably because vscode detects some background processes belongs to Continue are not terminated after users disable it. Not calling `.dispose()` on every `Disposable` might also be the cause. Anyway, If users don't click the "Restart Extensions" button after disabling Continue, the effects provided by those `Disposable` (not in `context.subscriptions`) will still be active and might cause some unexpected behaviors. However, all these will be fixed when users feel vscode is working weird and relaunch it.

owtaylor reviewed (changes requested):
This looks very good now - exactly how I'd expect it to be structured. Just one minor nit - otherwise good to go.

owtaylor commented:
Looks great, I'll go ahead and merge.
